### PR TITLE
KAFKA-6948 Change comparison to avoid overflow inconsistencies

### PR DESCRIPTION
**Pull request by thisthat**
Change timestamp comparison following what the Java documentation recommends to help preventing such errors.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
Thanks for the PR. Which java documentation are you referring to?

**Comment by thisthat** [comment]
@hachikuji thank you for the answer and sorry for the delay! 
The documentation that I am referring to the following: [link](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--)

File: clients/src/main/java/org/apache/kafka/clients/NetworkClientUtils.java [hunk 1]
```diff
@@ -65,7 +65,7 @@ public static boolean awaitReady(KafkaClient client, Node node, Time time, long
             return true;
 
         long attemptStartTime = time.milliseconds();
-        while (!client.isReady(node, attemptStartTime) && attemptStartTime < expiryTime) {
+        while (!client.isReady(node, attemptStartTime) && attemptStartTime - expiryTime < 0) {
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClientUtils.java:68]: Would it make sense to create a utility function for this given it's used in 9 places?

**Review by viktorsomogyi — COMMENTED** [review]
According to the javadocs of `System.nanoTime()` and `Syatem.currentTimeMillis()` overflows can happen only with the first one. How is it possible with the second?

**Comment by guozhangwang** [comment]
I'm okay with this change, and also I think it is better as a general coding style suggestion to use the proposed pattern whenever possible.

@thisthat could you rebase?

**Comment by thisthat** [comment]
@guozhangwang sure! Should I create an utility function as @viktorsomogyi suggested or just the rebase is sufficient?

**Comment by thisthat** [comment]
@viktorsomogyi the documentation presents the error only for the nano method call. However, the overflow can happen also with the millis call, e.g. KAFKA-4290.

**Comment by guozhangwang** [comment]
I think it is okay to just fix all the places: even if we introduce the util function, we cannot enforce future contributors to leverage on it. I think it is more practical to ask reviewers be more keen to such coding styles moving forward.

**Comment by thisthat** [comment]
@guozhangwang the rebase should be fine now

**Comment by guozhangwang** [comment]
@thisthat Could you rebase? Seems like a lot of recent commits were reverted in the diff file.

**Comment by thisthat** [comment]
@guozhangwang rebased!

**Comment by guozhangwang** [comment]
Hmm.. it still contains 20K+ diffs, if the branch is a bit messed you could consider creating a new branch out of current trunk.

**Comment by thisthat** [comment]
@guozhangwang agree! I will create a new PR and close this one 😉

**Pull request closed without merging**